sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,571 @@
+}
+
+EnqueueLibraryKernel(queue) {
+  queue.submit(parallel_for<ExternalKernel>(...));
+}
+```
+Such case can be optimized by bringing nesting into cache keys structure.
+Kernel set id can be found for each kernel using its name and OS module it is
+coming from. In presence of dynamic linking resulting program can be combined
+out of device images which come from different OS modules. So, it should be
+possible to find needed program by kernel name and any OS module that was
+involved in this program. The new mapping structure is:
+```
+{kernel name} =>
+  {OSModuleHandle, spec const, opts, dev} => program
+```
+I.e. each kernel name is mapped to a set of tuples that consists of OS module,
+spec constant values, JIT compiler options and device. Then concrete tuple is
+mapped to a program object. Several tuples can be mapped to a same program
+object, they are created during process of compilation and symbols resolution
+for concrete device image. When some program is made through linking of several
+programs created from device images that come from different OS modules,
+for each OS module in cache will be created a tuple with corresponding OS module
+id.
+Example of modified cache structure when dynamic linking is involved:
+```
+// Application
+// OSModule = 1
+
+SYCL_EXTERNAL void LibFunc();
+queue.submit(parallel_for<InternalKernel>( ... ));
+
+Q.submit([&](cl::sycl::handler &CGH) {
+  CGH.parallel_for([] { LibFunc(); });
+});
+
+EnqueueLibraryKernel(q);
+
+// Library
+// OSModule = 2
+
+SYCL_EXTERNAL LibFunc();
+
+EnqueueLibraryKernel(queue) {
+  queue.submit(parallel_for<ExternalKernel>(...));
+}
+
+Program cache will have the following structure:
+ "InternalKernel" =>
+   {1, ...} => program 1
+ "ExternalKernel" =>
+   {1, ...} => program 1
+   {2, ...} => program 1
+```
+However the library code will be compiled twice if kernel from the library
+was enqueued before kernels from the application, i.e. in such case:
+```
+// Application
+SYCL_EXTERNAL void LibFunc();
+
+EnqueueLibraryKernel(Q); // First, library code is compiled alone since it
+                         // doesn't have any dependencies
+// ...
+Q.submit([&](cl::sycl::handler &CGH) {
+  handler.parallel_for([] { LibFunc(); });  // Second, library code is compiled
+                                            // and linked together with code of
+                                            // the application
+});
+```
+
+The kernel caching mechanism is re-used without changes.
+
+##### Persistent cache
+
+The documented approach to persistent cache needs to be expanded in presence
+of dynamic linking support. One of the identifiers for built image hash is
+hash made out of device image used as input for the JIT compilation.
+In case when "main" image have imports information, device image hash should be
+created from all device images that are necessary to build it, i.e. hash out
+of "main" device image and set of images that define all
+symbols imported by "main" device image.
```

Comment:
That is a good point. With the current scheme we cannot guarantee persistent order of device images. We could actually sort device images by some feature (like size, set of defined symbols and etc.) before creating a hash string. But the algorithm of search itself doesn't prevent using of different images to resolve dependencies, since it is assumed that the same symbol defined in several device images of the same format should have the same definition. However when some other device image is chosen to resolve dependency, it won't be technically the same program as the result of linking.
So, does sorting of device images before creation of hash string sounds ok?

Reply:
> Though, using OS module ID (a mere number) as sorting feature is of more optimized way.

Do you mean OS module handle? Isn't just a pointer?

https://github.com/intel/llvm/blob/7d5ee05b5460c0df7090f32f59efa2fcb981d73d/sycl/include/CL/sycl/detail/os_util.hpp#L48

I'm not really sure that I understand how getting of OS module handle works, but can it change from run to run?

Reply:
> Do you mean OS module handle? Isn't just a pointer?

My bad, then.

Reply:
@s-kanaev , do you mind if I make this addition and apply other NIT comments from you in the follow-up patch? I think it will take very much time to get all the approvals again.

Reply:
I don't mind at all.